src/pkg/scan/rest/auth/auth.go

```diff
@@ -0,0 +1,54 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	authorization = "Authorization"
+	// Basic ...
+	Basic = "Basic"
+	// Bearer ...
+	Bearer = "Bearer"
+	// APIKey ...
+	APIKey = "X-ScannerAdapter-API-Key"
+)
+
+// Authorizer defines operation for authorizing the requests
+type Authorizer interface {
+	Authorize(req *http.Request) error
+}
+
+// GetAuthorizer is a factory method for getting an authorizer based on the given auth type
+func GetAuthorizer(auth, cred string) (Authorizer, error) {
```

Comment:
I do not think we need so many authorizers, such `apiKeyAuthorizer`, `basicAuthorizer` and `bearerAuthorizer` and there are many common code in these implements.

```go
type defaultAuthorizer struct {
    prefix string
    value string
}
```

`defaultAuthorizer` (maybe other name) can cover apikey, base, bearer scenarios.

For basic mode, the value can be encoded as base64 before the creating of the authorizer.

The validation of `cred` can be done in GetAuthorizer, currently the behave to validate the `cred` is not same in these three authorizers.

Reply:
The design idea is aligned with the `Auth` property in the registration model. With auth factory, it is easy to get one by that property. And explicitly declaring the different kinds of authorizer will more flexible and clear (Although some duplicated code generated).

Moreover, it is easier for API consumer (like the portal API ) to consume and mapping the data.

It is a design pattern, and I don't think your proposed merged authorizer is better than the current one.

For the cred validation, it should be done by the authorizer themselves. They may have different requirements. (At present, the no authorizer will not validate the cred).

Reply:
ok